### PR TITLE
GH-1301 control connective operator precedence

### PR DESCRIPTION
**Pull request by abhishekag03**
GitHub issue resolved: #1301  

Briefly describe the changes proposed in this PR:
Used parenthesize function in case of AND operator by calling it in the Operation method present in sparqlbuilder/constraint/Operation.java file.

---- 
PR Author Checklist: 

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [x] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

Note: we merge all feature pull requests using [squash and merge](https://help.github.com/en/github/administering-a-repository/about-merge-methods-on-github#squashing-your-merge-commits). See [RDF4J git merge strategy](https://rdf4j.org/documentation/developer/merge-strategy/) for more details.

### CODE REVIEW

**Comment by barthanssens** [comment]
perhaps @anmolsinghal98 could take a look at this ?

**Comment by anmolsinghal98** [comment]
Do you want me to work on the issue @barthanssens ?

**Comment by barthanssens** [comment]
Hi @anmolsinghal98 

I was wondering if perhaps you could take a look at the pull request abhishek created.
(the code and test looks good to me, but I'm not familiar with the SparqlBuilder, so help is appreciated)

**Comment by abhishekag03** [comment]
Hey, @anmolsinghal98 please let us know in case you need any clarifications related to the Pull Request submitted by us?

**Comment by anmolsinghal98** [comment]
Hey. Apologies but it would not be possible for me to review it.  The contributions we are making in rdf4j are part of a course project. Since both Abhishek and I are enrolled in the same course, my review would act as a conflict of interest and the course instructor has advised against it. Hope you understand @barthanssens

**Comment by barthanssens** [comment]
Ow, I didn't know that. 
Thank you all for choosing RDF4J as part of your course projects ! I'll try to look into it

File: core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/constraint/Operation.java [hunk 1]
```diff
@@ -23,7 +23,9 @@
 	Operation(SparqlOperator operator, int operandLimit) {
 		super(operator);
 		this.operandLimit = operandLimit;
-//		parenthesize();
+		if (operator.toString().equals("AND")) {
```
Comment (hmottestad) [inline core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/constraint/Operation.java:26]: Using toString() and then .equals() is typically very slow. Try to use the enums instead: ConnectiveOperator.AND

File: core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/examples/sparql11spec/Section6Test.java [hunk 2]
```diff
@@ -61,4 +61,19 @@ public void example_6_3() {
...
+
+		GraphPatternNotTriples pricePattern = GraphPatterns.and(x.has(ns.iri("price"), price))
+				.filter(Expressions.or(Expressions.lt(price, 20),
+						Expressions.and(Expressions.lt(price, 50), Expressions.gt(price, 30))))
+				.optional();
+
+		query.prefix(dc, ns).select(title, price).where(x.has(dc.iri("title"), title), pricePattern);
+		p();
+	}
+
```
Comment (hmottestad) [inline core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/examples/sparql11spec/Section6Test.java:78]: I don't see how this tests actually checks if the changes you've made are correct. The test still passes even if I comment out your change.

Comment (hmottestad) [reply core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/examples/sparql11spec/Section6Test.java:78]: Btw. I'm sorry the test code in general here is so poor. `p()` isn't particularly useful here (for those reading this comment it actually means System.out.println(query) ), and all the tests are stateful, even though there is no need to keep state between tests.

Comment (hmottestad) [reply core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/examples/sparql11spec/Section6Test.java:78]: Second btw. I've take another look at the tests within the `examples` package and realises that none of these test anything, ie. there are no asserts anywhere. When I remove these "tests" I end up with a 13% test coverage. This is very unfortunate, but in now way your fault @abhishekag03

Comment (hmottestad) [reply core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/examples/sparql11spec/Section6Test.java:78]: I propose this assert:

```
Assert.assertThat(query.getQueryString(), CoreMatchers.containsString("( ?price < 50 && ?price > 30 )"));
```

File: core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/constraint/Operation.java (hunk 1 above)
Comment (abhishekag03) [reply core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/constraint/Operation.java]: Thanks for the suggestion! Done the necessary changes.

File: core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/constraint/Operation.java [hunk 3]
```diff
@@ -23,7 +23,7 @@
 	Operation(SparqlOperator operator, int operandLimit) {
 		super(operator);
 		this.operandLimit = operandLimit;
-		if (operator == ConnectiveOperator.AND) {
+		if (operator == ConnectiveOperator.AND || operator == ConnectiveOperator.OR) {
```
Comment (abhishekag03) [inline core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/constraint/Operation.java:26]: We noticed that the Paranthesize function needs to be called for the OR operator as well. This is because for queries which have an OR inside an AND, the query output is not how it should be. It does not take into consideration that the OR part must be evaluated first. 
However, calling the Paranthesize function as we have done here is giving us extra parenthesis inside the filter query whenever we have boolean binary operators (AND/OR). Is that fine? If not, kindly help us with how we should remove it. Although considering it evaluation wise, it will always give the correct output. Just that the brackets there are not essential.

Comment (hmottestad) [reply core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/constraint/Operation.java:26]: I'm ok with the extra parenthesis. I think it would be hard to fix and it's not semantically incorrect, just superfluous.

**Comment by hmottestad** [comment]
LGTM. 

@jeenbroekstra what do you think?

File: core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/constraint/Operation.java (hunk 3 above)
Comment (hmottestad) [reply core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/constraint/Operation.java:26]: I am wondering if all connective operators need parenthesises around them? Do you know @abhishekag03 ?

Comment (abhishekag03) [reply core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/constraint/Operation.java:26]: Yeah, you're actually right! We need to include it for all Connective Operators. We will add the code for the same along with test cases for each operator. I'm just wondering if this should be done in this issue itself or shall we open a new issue and commit the changes there? Or can the issue heading be changed? Please advise what shall be done?

Comment (hmottestad) [reply core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/constraint/Operation.java:26]: It’ll be easier to do it in this PR and just update the title and description of both the issue and this PR.

Comment (abhishekag03) [reply core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/constraint/Operation.java]: Sure, we have done the same and pushed the code. Since the check had to be done for all connective operators, we have now used `instanceOf` instead of equality check. Kindly have a look and let us know if something needs to be changed.

File: core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/examples/sparql11spec/Section6Test.java [hunk 4]
```diff
@@ -61,4 +64,65 @@ public void example_6_3() {
...
+		Variable title = SparqlBuilder.var("title"), price = SparqlBuilder.var("price"), x = SparqlBuilder.var("x");
+
+		GraphPatternNotTriples pricePattern = GraphPatterns.and(x.has(ns.iri("price"), price))
+				.filter(Expressions.or(Expressions.lt(price, Expressions.subtract(Rdf.literalOf(20),
+						Expressions.multiply(Rdf.literalOf(2), Rdf.literalOf(5)))),
+						Expressions.lt(price, 50)))
+				.optional();
+
+		query.prefix(dc, ns).select(title, price).where(x.has(dc.iri("title"), title), pricePattern);
+		Assert.assertThat(query.getQueryString(), CoreMatchers.containsString("( 20 - ( 2 * 5 ) )"));
```
Comment (hmottestad) [inline core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/examples/sparql11spec/Section6Test.java:110]: 20-2*5 == 20-(2*5)

Could you instead test for (20-2)*5 since this is different from 20-2*5?

Comment (abhishekag03) [reply core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/examples/sparql11spec/Section6Test.java:110]: Sure. Added the same and pushed it. Also, shall we add all these test cases to the file Section6Test.java only or should there be a separate file for these test cases? I couldn't understand how the Test cases are divided into the various SectionTest files.

Comment (hmottestad) [reply core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/examples/sparql11spec/Section6Test.java:110]: As I stated earlier. There are actually almost no tests for the SparqlBuilder (ie. no tests will fail if I change random things in the code). I think adding the tests like you have done here should be fine. But @jeenbroekstra should take a look to see if he agrees. Jeen is the project manager for RDF4J.

Comment (hmottestad) [reply core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/examples/sparql11spec/Section6Test.java:110]: Also, thank you @abhishekag03!

Comment (abhishekag03) [reply core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/examples/sparql11spec/Section6Test.java:110]: Thank you for the valuable suggestions and prompt responses @hmottestad. 
Will wait for @jeenbroekstra to have a look at the changes made so that this issue can be closed.

**Comment by jeenbroekstra** [comment]
@ahmadbanijamali with respect to the tests for this: as @hmottestad also mentioned, the tests in package `org.eclipse.rdf4j.sparqlbuilder.examples` are not actually real tests: all they do is create a query and print it out. The different sections refer to the sections of the [SPARQL 1.1 Query Language W3C Recommendation](https://www.w3.org/TR/sparql11-query/) - the examples recreate the queries used in the various sections there. 

I would suggest to not edit any of the code existing in `org.eclipse.rdf4j.sparqlbuilder.examples`, and instead introduce your tests as part of other test classes more closely linked to the actual structure and function of what you're testing. In this case, since you are testing the working of 'or' and 'and', which are both expressions, I'd add some a few very specific tests to `org.eclipse.rdf4j.sparqlbuilder.constraint.ExpressionsTest`, and the more general tests (which use the entire SparqlBuilder to create a full query) to `org.eclipse.rdf4j.sparqlbuilder.core.SparqlBuilderTest` (doesn't exist yet, feel free to create).

File: core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/examples/sparql11spec/Section6Test.java (hunk 4 above)
Comment (jeenbroekstra) [reply core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/examples/sparql11spec/Section6Test.java:110]: I added a comment on the PR about this, but just to repeat here: I suggest you move these tests out of these `SectionTest` classes completely, and instead introduce a test class `org.eclipse.rdf4j.sparqlbuilder.core.SparqlBuilderTest` where you can add these cases. I'd prefer not to edit the code in the package `org.eclipse.rdf4j.sparqlbuilder.examples` - they are meant to be examples rather than actual code (I don't know if we want to keep them that way, but for now I'd like our actual tests to be separate from it).

**Comment by abhishekag03** [comment]
Thank you @jeenbroekstra @barthanssens @hmottestad. We have made the required changes and pushed the code. Kindly review it once. Please let us know if something is still remaining.
P.S: Kindly update the description and title of the issue to include precedence order of all the connective operators.

File: core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/core/SparqlBuilderTest.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
+/*******************************************************************************
+ * Copyright (c) ${year} Eclipse RDF4J contributors.
```
Comment (jeenbroekstra) [inline core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/core/SparqlBuilderTest.java:2]: Looks like your editor forgot to substitute `${year}` with 2020 :)

File: core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/core/SparqlBuilderTest.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+import static org.eclipse.rdf4j.sparqlbuilder.rdf.Rdf.iri;
+
+import org.eclipse.rdf4j.model.vocabulary.DC;
+import org.eclipse.rdf4j.sparqlbuilder.constraint.Expressions;
+import org.eclipse.rdf4j.sparqlbuilder.core.query.Queries;
+import org.eclipse.rdf4j.sparqlbuilder.core.query.SelectQuery;
+import org.eclipse.rdf4j.sparqlbuilder.graphpattern.GraphPatternNotTriples;
+import org.eclipse.rdf4j.sparqlbuilder.graphpattern.GraphPatterns;
+import org.eclipse.rdf4j.sparqlbuilder.rdf.Rdf;
+import org.hamcrest.CoreMatchers;
```
Comment (jeenbroekstra) [inline core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/core/SparqlBuilderTest.java:20]: We actually prefer using AssertJ rather than Hamcrest, but no big deal, you can leave as-is.

File: core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/core/SparqlBuilderTest.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+	public void testLogicalOperatorPrecedence1() {
+		Prefix dc = SparqlBuilder.prefix("dc", iri(DC_NS)), ns = SparqlBuilder.prefix("ns", iri(EXAMPLE_ORG_NS));
+		Variable title = SparqlBuilder.var("title"), price = SparqlBuilder.var("price"), x = SparqlBuilder.var("x");
+
+		GraphPatternNotTriples pricePattern = GraphPatterns.and(x.has(ns.iri("price"), price))
+				.filter(Expressions.or(Expressions.lt(price, Rdf.literalOf(20)),
+						Expressions.and(Expressions.lt(price, Rdf.literalOf(50)),
+								Expressions.gt(price, Rdf.literalOf(30)))))
+				.optional();
+		System.out.println(query);
```
Comment (jeenbroekstra) [inline core/sparqlbuilder/src/test/java/org/eclipse/rdf4j/sparqlbuilder/core/SparqlBuilderTest.java:49]: Could you remove the System.out.println ? Just so there is less output to scroll through when running these tests.

**Comment by abhishekag03** [comment]
We have made all the suggested changes and pushed the final code. 
It was great working with you @jeenbroekstra @hmottestad @barthanssens.
It was a great first experience for us. Looking forward to more open source contributions in the future.

Thanks a lot!

**Comment by hmottestad** [comment]
LGTM
